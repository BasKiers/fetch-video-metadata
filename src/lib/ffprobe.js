// @flow
/*
 * Based on the ffprobe npm package
 * ref: https://github.com/eugeneware/ffprobe/blob/master/index.js
 */

import JSONStream from 'JSONStream';
import bl from 'bl';
import { spawn } from 'child_process';
import path from 'path';

const ffprobePath = path.join(__dirname, '../../bin/ffprobe');

type ProbeResponse = ProbeResult | ProbeError;

export type ProbeError = {
	error: {
		code: number,
		string: string,
	}
}

export type ProbeResult = {
	streams: Array<VideoStreamData | AudioStreamData>,
	format: FormatData
}

export type VideoStreamData = {
	codec_type: 'video',
	bit_rate: ?string,
	r_frame_rate: ?string,
	coded_width: ?number,
	coded_height: ?number
}

export type AudioStreamData = {
	codec_type: 'audio',
	bit_rate: ?string,
	channel_layout: ?string,
	channels: ?number,
	sample_rate: ?string
}

export type FormatData = {
	duration: ?string,
	bit_rate: ?string
}

export default function ffprobe(filePath: string, opts: string[] = []): Promise<ProbeResult> {
	const params = [...opts, '-show_streams', '-show_format', '-show_error', '-print_format', 'json', filePath];

	return new Promise((resolve, reject) => {
		let info: ProbeResponse;
		let stderr: string;
		const ffprobe = spawn(ffprobePath, params);

		ffprobe.once('close', function (code: number) {
			if (code > 1 || code === 1 && !info) {
				const err = (stderr || '').split('\n').filter(Boolean).pop();
				return reject(new Error(err));
			}
			if (info.error) {
				return reject(new Error(info.error.string || 'Could not get metadata for given file'));
			}
			resolve(info);
		});

		ffprobe.stderr.pipe(bl((err: Error, data: any) => {
			stderr = data.toString();
		}));

		ffprobe.stdout
			.pipe(JSONStream.parse())
			.once('data', function (data: ProbeResponse) {
				info = data;
			});
	});
}
