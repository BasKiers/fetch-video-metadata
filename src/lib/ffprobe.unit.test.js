import ffprobe from './ffprobe';
import fs from 'fs';
import path from 'path';
import { promisify } from 'util';

describe('Test ffprobe usage', () => {

	test('ffprobe binary should be present and executable', () => {
		return promisify(fs.access)(path.join(__dirname, '../../bin/ffprobe'), fs.constants.X_OK)
	});

	test('returns an error on an invalid file path', () => {
		return expect(ffprobe('invalid')).rejects.toThrow('No such file or directory');
	});

	test('returns an error on an invalid url', () => {
		return expect(ffprobe('http://www.invalidurl123456789.co.uk')).rejects.toThrow('Input/output error');
	});

	test('should return an error on a valid url with an invalid media file', () => {
		return expect(ffprobe('http://www.google.com')).rejects.toThrow('Invalid data found when processing input');
	});

	test('returns a valid response on a valid mp4 media file url', async () => {
		const metadata = await ffprobe('https://s3-eu-west-1.amazonaws.com/tradecast-development-test/sample-video/tradecast.mp4');

		expect(metadata).toEqual(
			expect.objectContaining({
				format: expect.objectContaining({
					duration: expect.any(String),
					bit_rate: expect.any(String),
				}),
				streams: expect.arrayContaining([
					expect.objectContaining({
						codec_type: 'video',
						bit_rate: expect.any(String),
						r_frame_rate: expect.any(String),
						coded_width: expect.any(Number),
						coded_height: expect.any(Number),
					}),
					expect.objectContaining({
						codec_type: 'audio',
						bit_rate: expect.any(String),
						channel_layout: expect.any(String),
						channels: expect.any(Number),
						sample_rate: expect.any(String),
					}),
				]),
			})
		);

		expect(metadata).toMatchSnapshot();
	});

	test('returns a valid response on a valid png media file url', async () => {
		const metadata = await ffprobe('http://tradecast.tv/img/tradecast-logo.png');

		expect(metadata).toEqual(
			expect.objectContaining({
				format: expect.objectContaining({
					// duration: expect.any(String),
					// bit_rate: expect.any(String),
				}),
				streams: expect.arrayContaining([
					expect.objectContaining({
						codec_type: 'video',
						// bit_rate: expect.any(String),
						r_frame_rate: expect.any(String),
						coded_width: expect.any(Number),
						coded_height: expect.any(Number),
					}),
				]),
			})
		);

		return expect(metadata).toMatchSnapshot();
	});

});