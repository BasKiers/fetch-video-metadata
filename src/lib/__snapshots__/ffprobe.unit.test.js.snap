// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ffprobe usage returns a valid response on a valid mp4 media file url 1`] = `
Object {
  "format": Object {
    "bit_rate": "1334445",
    "duration": "90.001000",
    "filename": "https://s3-eu-west-1.amazonaws.com/tradecast-development-test/sample-video/tradecast.mp4",
    "format_long_name": "QuickTime / MOV",
    "format_name": "mov,mp4,m4a,3gp,3g2,mj2",
    "nb_programs": 0,
    "nb_streams": 2,
    "probe_score": 100,
    "size": "15012674",
    "start_time": "0.000000",
    "tags": Object {
      "compatible_brands": "isomiso2avc1mp41",
      "encoder": "Lavf57.25.100",
      "episode_sort": "1",
      "hd_video": "0",
      "major_brand": "isom",
      "media_type": "9",
      "minor_version": "512",
      "season_number": "1",
    },
  },
  "streams": Array [
    Object {
      "avg_frame_rate": "0/0",
      "bit_rate": "160000",
      "bits_per_sample": 0,
      "channel_layout": "stereo",
      "channels": 2,
      "codec_long_name": "AAC (Advanced Audio Coding)",
      "codec_name": "aac",
      "codec_tag": "0x6134706d",
      "codec_tag_string": "mp4a",
      "codec_time_base": "1/44100",
      "codec_type": "audio",
      "disposition": Object {
        "attached_pic": 0,
        "clean_effects": 0,
        "comment": 0,
        "default": 1,
        "dub": 0,
        "forced": 0,
        "hearing_impaired": 0,
        "karaoke": 0,
        "lyrics": 0,
        "original": 0,
        "timed_thumbnails": 0,
        "visual_impaired": 0,
      },
      "duration": "89.953991",
      "duration_ts": 3966971,
      "index": 0,
      "max_bit_rate": "160000",
      "nb_frames": "3876",
      "profile": "LC",
      "r_frame_rate": "0/0",
      "sample_fmt": "fltp",
      "sample_rate": "44100",
      "start_pts": 0,
      "start_time": "0.000000",
      "tags": Object {
        "handler_name": "SoundHandler",
        "language": "eng",
      },
      "time_base": "1/44100",
    },
    Object {
      "avg_frame_rate": "25/1",
      "bit_rate": "1167864",
      "bits_per_raw_sample": "8",
      "chroma_location": "left",
      "codec_long_name": "H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10",
      "codec_name": "h264",
      "codec_tag": "0x31637661",
      "codec_tag_string": "avc1",
      "codec_time_base": "1/50",
      "codec_type": "video",
      "coded_height": 720,
      "coded_width": 1280,
      "display_aspect_ratio": "16:9",
      "disposition": Object {
        "attached_pic": 0,
        "clean_effects": 0,
        "comment": 0,
        "default": 1,
        "dub": 0,
        "forced": 0,
        "hearing_impaired": 0,
        "karaoke": 0,
        "lyrics": 0,
        "original": 0,
        "timed_thumbnails": 0,
        "visual_impaired": 0,
      },
      "duration": "90.000000",
      "duration_ts": 1152000,
      "has_b_frames": 2,
      "height": 720,
      "index": 1,
      "is_avc": "true",
      "level": 31,
      "nal_length_size": "4",
      "nb_frames": "2250",
      "pix_fmt": "yuv420p",
      "profile": "Main",
      "r_frame_rate": "25/1",
      "refs": 1,
      "sample_aspect_ratio": "1:1",
      "start_pts": 0,
      "start_time": "0.000000",
      "tags": Object {
        "handler_name": "VideoHandler",
        "language": "eng",
      },
      "time_base": "1/12800",
      "width": 1280,
    },
  ],
}
`;

exports[`Test ffprobe usage returns a valid response on a valid png media file url 1`] = `
Object {
  "format": Object {
    "filename": "http://tradecast.tv/img/tradecast-logo.png",
    "format_long_name": "piped png sequence",
    "format_name": "png_pipe",
    "nb_programs": 0,
    "nb_streams": 1,
    "probe_score": 99,
    "size": "3792",
  },
  "streams": Array [
    Object {
      "avg_frame_rate": "0/0",
      "codec_long_name": "PNG (Portable Network Graphics) image",
      "codec_name": "png",
      "codec_tag": "0x0000",
      "codec_tag_string": "[0][0][0][0]",
      "codec_time_base": "0/1",
      "codec_type": "video",
      "coded_height": 80,
      "coded_width": 320,
      "color_range": "pc",
      "display_aspect_ratio": "0:1",
      "disposition": Object {
        "attached_pic": 0,
        "clean_effects": 0,
        "comment": 0,
        "default": 0,
        "dub": 0,
        "forced": 0,
        "hearing_impaired": 0,
        "karaoke": 0,
        "lyrics": 0,
        "original": 0,
        "timed_thumbnails": 0,
        "visual_impaired": 0,
      },
      "has_b_frames": 0,
      "height": 80,
      "index": 0,
      "level": -99,
      "pix_fmt": "pal8",
      "r_frame_rate": "25/1",
      "refs": 1,
      "sample_aspect_ratio": "0:1",
      "time_base": "1/25",
      "width": 320,
    },
  ],
}
`;
